receiver_test.go

```diff
@@ -2,137 +2,156 @@ package amqp
+	}
+	// wait for the link to unpause as credit should now be available
+	ctx, cancel = context.WithTimeout(context.Background(), 1*time.Second)
+	for {
+		if atomic.LoadUint32(&r.link.Paused) == 0 {
+			cancel()
+			break
+		} else if err := ctx.Err(); err != nil {
+			t.Fatal(err)
+		}
 	}
-	if err := r.HandleMessage(context.TODO(), doNothing); err != nil {
-		t.Errorf("HandleMessage() error = %v", err)
+	// link credit should be 1
+	if c := r.link.linkCredit; c != 1 {
+		t.Fatalf("unexpected link credit %d", c)
 	}
-
-	if len(r.dispositions) != 0 {
-		t.Errorf("the message should not have triggered a disposition")
+	// subsequent dispositions should have no effect
+	if err := r.AcceptMessage(context.Background(), msg); err != nil {
+		t.Fatal(err)
 	}
 }
 
-func TestReceiver_HandleMessageModeSecond_DontDispose(t *testing.T) {
-	r := &Receiver{
-		link:         makeLink(ModeSecond),
-		batching:     true, // allows to  avoid making the outgoing call on dispostion
-		dispositions: make(chan messageDisposition, 2),
-	}
-	msg := makeMessage(ModeSecond)
-	r.link.Messages <- msg
-	r.link.addUnsettled(&msg)
-	if err := r.HandleMessage(context.TODO(), doNothing); err != nil {
-		t.Errorf("HandleMessage() error = %v", err)
-	}
-	if len(r.dispositions) != 0 {
-		t.Errorf("it is up to the message handler to settle messages")
-	}
-	if r.link.countUnsettled() == 0 {
-		t.Errorf("the message should still be tracked until settled")
-	}
-	// ensure channel wasn't closed
-	select {
-	case _, ok := <-msg.doneSignal:
-		if !ok {
-			t.Fatal("unexpected closing of doneSignal")
+func TestReceive_ModeSecond(t *testing.T) {
+	mock := mocks.NewConnection()
+	// create sequence of AMQP handshake frames
+	mock.AppendProtoHeader(mocks.ProtoAMQP)
+	assert.NoError(t, mock.AppendPerformOpen("test"))
+	client, err := New(mock)
+	if err != nil {
+		t.Fatal(err)
+	}
+	assert.NoError(t, mock.AppendPerformBegin(0))
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	session, err := client.NewSession(ctx)
+	cancel()
+	if err != nil {
+		t.Fatal(err)
+	}
+	const linkName = "test"
+	const linkHandle = 0
+	assert.NoError(t, mock.AppendReceiverAttach(linkName, linkHandle, ModeSecond))
+	ctx, cancel = context.WithTimeout(context.Background(), 1*time.Second)
+	r, err := session.NewReceiver(ctx, LinkName(linkName), LinkReceiverSettle(ModeSecond))
+	cancel()
+	if err != nil {
+		t.Fatal(err)
+	}
+	deliveryID := uint32(1)
+	// create a mock message for retrieval
+	assert.NoError(t, mock.AppendTransfer(linkHandle, deliveryID, []byte("hello")))
+	ctx, cancel = context.WithTimeout(context.Background(), 1*time.Second)
+	msg, err := r.Receive(ctx)
+	cancel()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if c := r.link.countUnsettled(); c != 1 {
+		t.Fatalf("unexpected unsettled count %d", c)
+	}
+	// wait for the link to pause as we've consumed all available credit
+	ctx, cancel = context.WithTimeout(context.Background(), 1*time.Second)
```

Comment:
btw, this is one of those instances I'm thinking something like https://github.com/Azure/go-amqp/issues/71 would be useful (do only a non-blocking select off the received messages channel).